File: 40.组合总和2/combinationSum2.go
package problem0040

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	var (
		result    [][]int
		backTrack func(path []int, target int, index int) (shouldGoOn bool)
	)
	backTrack = func(path []int, target int, index int) (shouldGoOn bool) {
		if target == 0 {
			temp := make([]int, len(path))
			copy(temp, path)
			result = append(result, temp)
			return true
		}
		if target < 0 {
			return false
		}
		for i := index; i < len(candidates); i++ {
			if i > index && candidates[i] == candidates[i-1] {
				continue
			}
			path = append(path, candidates[i])
			if !backTrack(path, target-candidates[i], i+1) {
				break
			}
			path = path[:len(path)-1]
		}
		return true
	}
	sort.Ints(candidates)
	backTrack([]int{}, target, 0)
	return result
}

File: 40.组合总和2/combinationSum2.js
function combinationSum2(candidates, target) {
    // 排序
    candidates.sort((a, b) => a < b ? -1 : 1)
    const result = []
    const backTrack = (combine, target, index) => {
        if (target < 0) {
            return false
        }
        if (target === 0) {
            result.push(combine)
            return true
        }
        for (let i = index; i < candidates.length; i++) {
            // 剪枝，如果相邻两个元素相等，跳过
            if (i > index && candidates[i] === candidates[i - 1]) {
                continue
            }
            combine.push(candidates[i])
            // 注意这里的引用
            if (!backTrack([...combine], target - candidates[i], i + 1)) {
                break
            }
            combine.pop()
        }
        return true
    }
    backTrack([], target, 0)
    return result
}